// @just-every/ensemble/dist/cjs/types/errors.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigurationError = exports.ModelNotFoundError = exports.QuotaExceededError = exports.PauseAbortError = exports.AbortError = exports.ToolExecutionError = exports.ProviderError = exports.EnsembleError = void 0;
class EnsembleError extends Error {
    code;
    recoverable;
    details;
    constructor(message, code, recoverable = false, details) {
        super(message);
        this.code = code;
        this.recoverable = recoverable;
        this.details = details;
        this.name = 'EnsembleError';
    }
}
exports.EnsembleError = EnsembleError;
class ProviderError extends EnsembleError {
    provider;
    constructor(provider, message, code, recoverable = false, details) {
        super(message, code, recoverable, details);
        this.provider = provider;
        this.name = 'ProviderError';
    }
}
exports.ProviderError = ProviderError;
class ToolExecutionError extends EnsembleError {
    toolName;
    constructor(toolName, message, code = 'TOOL_EXECUTION_ERROR', recoverable = true, details) {
        super(message, code, recoverable, details);
        this.toolName = toolName;
        this.name = 'ToolExecutionError';
    }
}
exports.ToolExecutionError = ToolExecutionError;
class AbortError extends EnsembleError {
    constructor(message = 'Operation aborted') {
        super(message, 'ABORT_ERROR', false);
        this.name = 'AbortError';
    }
}
exports.AbortError = AbortError;
class PauseAbortError extends AbortError {
    constructor() {
        super('Operation aborted while waiting for pause');
        this.name = 'PauseAbortError';
    }
}
exports.PauseAbortError = PauseAbortError;
class QuotaExceededError extends EnsembleError {
    provider;
    model;
    constructor(provider, model, message, details) {
        super(message, 'QUOTA_EXCEEDED', false, details);
        this.provider = provider;
        this.model = model;
        this.name = 'QuotaExceededError';
    }
}
exports.QuotaExceededError = QuotaExceededError;
class ModelNotFoundError extends EnsembleError {
    model;
    constructor(model, message) {
        super(message || `Model '${model}' not found or not available`, 'MODEL_NOT_FOUND', false);
        this.model = model;
        this.name = 'ModelNotFoundError';
    }
}
exports.ModelNotFoundError = ModelNotFoundError;
class ConfigurationError extends EnsembleError {
    constructor(message, details) {
        super(message, 'CONFIGURATION_ERROR', false, details);
        this.name = 'ConfigurationError';
    }
}
exports.ConfigurationError = ConfigurationError;
