// @just-every/crawl/dist/cache/disk.js
import { createHash } from 'crypto';
import { mkdir, readFile, writeFile, access } from 'fs/promises';
import { join } from 'path';
export class DiskCache {
    cacheDir;
    cacheEnabled = true;
    constructor(cacheDir = '.cache') {
        this.cacheDir = cacheDir;
    }
    async init() {
        try {
            await mkdir(this.cacheDir, { recursive: true });
        }
        catch (error) {
            // If we can't create the cache directory, disable caching
            // Only show warning in debug mode or when explicitly requested
            if (process.env.DEBUG || process.env.CRAWL_DEBUG) {
                console.warn(`Cache directory creation failed: ${error instanceof Error ? error.message : 'Unknown error'}. Caching will be disabled.`);
            }
            this.cacheEnabled = false;
        }
    }
    getCacheKey(url) {
        return createHash('sha256').update(url).digest('hex');
    }
    getCachePath(url) {
        const key = this.getCacheKey(url);
        return join(this.cacheDir, `${key}.json`);
    }
    async has(url) {
        if (!this.cacheEnabled)
            return false;
        try {
            await access(this.getCachePath(url));
            return true;
        }
        catch {
            return false;
        }
    }
    async get(url) {
        if (!this.cacheEnabled)
            return null;
        try {
            const path = this.getCachePath(url);
            const data = await readFile(path, 'utf-8');
            return JSON.parse(data);
        }
        catch {
            return null;
        }
    }
    async put(url, markdown, title) {
        if (!this.cacheEnabled)
            return;
        const entry = {
            url,
            markdown,
            timestamp: Date.now(),
            title,
        };
        const path = this.getCachePath(url);
        try {
            await writeFile(path, JSON.stringify(entry, null, 2));
        }
        catch (error) {
            // Only show warning in debug mode
            if (process.env.DEBUG || process.env.CRAWL_DEBUG) {
                console.warn(`Failed to write to cache: ${error instanceof Error ? error.message : 'Unknown error'}`);
            }
        }
    }
    async getAge(url) {
        const entry = await this.get(url);
        if (!entry)
            return null;
        return Date.now() - entry.timestamp;
    }
}
