// @just-every/ensemble/dist/cjs/utils/external_models.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerExternalModel = registerExternalModel;
exports.getExternalModel = getExternalModel;
exports.getExternalProvider = getExternalProvider;
exports.isExternalModel = isExternalModel;
exports.clearExternalRegistrations = clearExternalRegistrations;
exports.overrideModelClass = overrideModelClass;
exports.getModelClassOverride = getModelClassOverride;
const externalModels = new Map();
const externalProviders = new Map();
const modelClassOverrides = new Map();
function registerExternalModel(model, provider) {
    const modelId = model.id;
    const modelWithDefaults = {
        ...model,
        cost: model.cost || {
            input_per_million: 0,
            output_per_million: 0,
        },
        features: model.features || {
            context_length: 1000000,
            tool_use: true,
            streaming: true,
            json_output: true,
            input_modality: ['text', 'image', 'audio', 'video'],
            output_modality: ['text'],
            max_output_tokens: 100000,
        },
    };
    externalModels.set(modelId, modelWithDefaults);
    if (!externalProviders.has(model.provider)) {
        externalProviders.set(model.provider, provider);
    }
    console.log(`[Ensemble] Registered external model: ${modelId} with provider: ${model.provider}`);
}
function getExternalModel(modelId) {
    return externalModels.get(modelId);
}
function getExternalProvider(providerId) {
    return externalProviders.get(providerId);
}
function isExternalModel(modelId) {
    return externalModels.has(modelId);
}
function clearExternalRegistrations() {
    externalModels.clear();
    externalProviders.clear();
    modelClassOverrides.clear();
}
function overrideModelClass(className, modelClass) {
    modelClassOverrides.set(className, modelClass);
    console.log(`[Ensemble] Overrode model class: ${className} with models: ${modelClass.models?.join(', ')}`);
}
function getModelClassOverride(className) {
    return modelClassOverrides.get(className);
}
