// @just-every/ensemble/dist/cjs/utils/pause_controller.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPauseController = getPauseController;
exports.isPaused = isPaused;
exports.pause = pause;
exports.resume = resume;
exports.waitWhilePaused = waitWhilePaused;
const event_emitter_js_1 = require("./event_emitter.cjs");
class PauseControllerImpl extends event_emitter_js_1.EventEmitter {
    _isPaused = false;
    isPaused() {
        return this._isPaused;
    }
    pause() {
        if (!this._isPaused) {
            this._isPaused = true;
            this.emit('paused');
            console.log('[PauseController] System paused');
        }
    }
    resume() {
        if (this._isPaused) {
            this._isPaused = false;
            this.emit('resumed');
            console.log('[PauseController] System resumed');
        }
    }
    async waitWhilePaused(checkInterval = 100, abortSignal) {
        while (this._isPaused && !abortSignal?.aborted) {
            await new Promise(resolve => setTimeout(resolve, checkInterval));
        }
        if (abortSignal?.aborted) {
            const { PauseAbortError } = await Promise.resolve().then(() => __importStar(require("../types/errors.cjs")));
            throw new PauseAbortError();
        }
    }
}
let pauseControllerInstance = null;
function getPauseController() {
    if (!pauseControllerInstance) {
        pauseControllerInstance = new PauseControllerImpl();
    }
    return pauseControllerInstance;
}
function isPaused() {
    return getPauseController().isPaused();
}
function pause() {
    getPauseController().pause();
}
function resume() {
    getPauseController().resume();
}
async function waitWhilePaused(checkInterval, abortSignal) {
    return getPauseController().waitWhilePaused(checkInterval, abortSignal);
}
