// @just-every/ensemble/dist/cjs/core/ensemble_image.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ensembleImage = ensembleImage;
const node_crypto_1 = require("node:crypto");
const model_provider_js_1 = require("../model_providers/model_provider.cjs");
function ensembleImage(prompt, agent, options = {}) {
    const run = async () => {
        const model = await (0, model_provider_js_1.getModelFromAgent)(agent, 'image_generation');
        const provider = (0, model_provider_js_1.getModelProvider)(model);
        if (!provider.createImage)
            throw new Error(`Provider for model ${model} does not support image generation`);
        return provider.createImage(prompt, model, agent, options);
    };
    if (!options.stream) {
        return run();
    }
    const self = async function* () {
        const request_id = options.request_id || (0, node_crypto_1.randomUUID)();
        options = { ...options, request_id };
        const { costTracker } = await Promise.resolve().then(() => __importStar(require("../utils/cost_tracker.cjs")));
        yield { type: 'image_start', request_id, timestamp: new Date().toISOString() };
        const handler = (usage) => {
            if (usage?.request_id === request_id) {
                const ev = { type: 'cost_update', usage, request_id, timestamp: new Date().toISOString() };
                iterator.push(ev);
            }
        };
        const iterator = {
            queue: [],
            push(e) { this.queue.push(e); },
        };
        costTracker.onAddUsage(handler);
        try {
            const images = await run();
            while (iterator.queue.length)
                yield iterator.queue.shift();
            let idx = 0;
            for (const img of images) {
                const isUrl = typeof img === 'string' && /^https?:\/\//i.test(img);
                const ev = {
                    type: 'file_complete',
                    request_id,
                    message_id: `${request_id}_img_${idx}`,
                    data_format: isUrl ? 'url' : 'base64',
                    data: img,
                    timestamp: new Date().toISOString(),
                };
                yield ev;
                while (iterator.queue.length)
                    yield iterator.queue.shift();
                idx++;
            }
            yield { type: 'image_complete', request_id, timestamp: new Date().toISOString() };
            while (iterator.queue.length)
                yield iterator.queue.shift();
        }
        catch (error) {
            yield { type: 'error', request_id, error: String(error?.message || error) };
        }
        finally {
            costTracker.offAddUsage(handler);
        }
    };
    return self();
}
